# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

#导入数据分析所需要的库，并通过Pandas的 read_csv 函数，将原始数据文件"e_commerce.csv"里的数据内容，解析为
DataFrame，并赋值给变量original_data。

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")
original_data.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

抽取若干数据，检查一下是否有结构性问题

In [3]:
original_data.sample(10)

InvoiceNo StockCode                        Description  Quantity  \
271593    560710      POST                            POSTAGE         1   
236189    557744     22720  SET OF 3 CAKE TINS PANTRY DESIGN          3   
511108    579471     20718          RED RETROSPOT SHOPPER BAG         3   
335428    566286     22949              36 DOILIES DOLLY GIRL        12   
167598    550995    85132C      CHARLIE AND LOLA FIGURES TINS        12   
433529    573927     23126          FELTCRAFT GIRL AMELIE KIT         4   
2775      536592     22631           CIRCUS PARADE LUNCH BOX          1   
482862    577485    85049E          SCANDINAVIAN REDS RIBBONS         2   
456854    575725     79321                      CHILLI LIGHTS        72   
145271    548893    46776B       WOVEN BERRIES CUSHION COVER          8   

             InvoiceDate  UnitPrice  CustomerID         Country  
271593   7/20/2011 12:51      15.00     14646.0     Netherlands  
236189   6/22/2011 12:09       4.95     15194.0  United Kingdom  
511108  11/29/2011 15:08       3.29         NaN  United Kingdom  
335428   9/11/2011 14:58       1.45     14057.0  United Kingdom  
167598    4/26/2011 8:33       1.95     16722.0  United Kingdom  
433529    11/2/2011 9:59       4.95     13890.0  United Kingdom  
2775     12/1/2010 17:06       4.21         NaN  United Kingdom  
482862  11/20/2011 11:56       1.25     16360.0  United Kingdom  
456854  11/10/2011 18:46       4.95     13098.0  United Kingdom  
145271    4/4/2011 15:54       4.13         NaN  United Kingdom

In [4]:
original_data.sample(10)

InvoiceNo StockCode                        Description  Quantity  \
537243    581238    85014b             RED RETROSPOT UMBRELLA         1   
222168    556365     84380  SET OF 3 BUTTERFLY COOKIE CUTTERS         3   
209158    555162     22431         WATERING CAN BLUE ELEPHANT         6   
528112    580729     21035    SET/2 RED RETROSPOT TEA TOWELS          1   
388498    570442     23198     PANTRY MAGNETIC  SHOPPING LIST         2   
17659     537772     22560         TRADITIONAL MODELLING CLAY         3   
292963    562573     22209      WOOD STAMP SET HAPPY BIRTHDAY        12   
356215    568047     23295   SET OF 12 MINI LOAF BAKING CASES        12   
369806    569103     23427             STOOL HOME SWEET HOME          1   
119169    546530     22720  SET OF 3 CAKE TINS PANTRY DESIGN         72   

             InvoiceDate  UnitPrice  CustomerID         Country  
537243   12/8/2011 10:53      12.46         NaN  United Kingdom  
222168   6/10/2011 11:44       1.25     16370.0  United Kingdom  
209158    6/1/2011 10:15       1.95     12473.0         Germany  
528112   12/5/2011 17:24       6.63         NaN  United Kingdom  
388498  10/10/2011 14:37       1.45     16549.0  United Kingdom  
17659    12/8/2010 12:35       1.25     18043.0  United Kingdom  
292963    8/7/2011 12:40       0.83     13248.0  United Kingdom  
356215   9/23/2011 12:27       0.83     15075.0  United Kingdom  
369806   9/30/2011 12:30      12.50     16133.0  United Kingdom  
119169   3/14/2011 13:25       4.95     17404.0          Sweden

从抽样的10行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商
品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


此时发现物品行数总共有541909条，其中Description和CustomerID出现了缺失数据的情况。InvoiceData为日期显示为字符串应当更改为日期，CustomerID为用户编号显示为浮点数，应该更改为字符串

### 评估缺失数据

根据此代码 original_data["Description"].isnull()，挑选出Description所在列的空缺值，利用其返回的布尔值进行索引，筛选出Descripton的缺失行

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

观察发现对于Description空缺的数据，unitprice在这些表格里均为0，所以猜想缺失数据的单价均为0，
不妨在进行一个条件& (original_data["UnitPrice"] != 0筛选出，缺少产品名称，但是单价不为0的情况

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0) ]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

结果显示表现出Description空缺时，单价unitprice确实均为0，由此根据数据分析目的，这俩同时缺失，说明对数据分析结果无法提供有效参考，应当在后续删除。

同理，接下来分析CustomID的缺失情况

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

我们可以看到，CustomerID缺失的时候，其他数据并不是全存在缺失的情况，并且CustomerID的值不会影响到我们对畅销物品的分析，所以可以保留。

### 评估重复数据

根据数据变量的含义来看，虽然InvoiceNo、stockcode 和 customerID 都是唯一标识符，但一次交易可能包含多件商
品，因此 InvoiceNo 可以存在重复，不同交易可以包含同件商品，因此 stockcode 可以存在重复，顾客可以进行多次交易或
下单多个商品，因此 CustomerID 也可以存在重复

### 评估不一致数据

我们可以对contry进行不一致数据评估，检测是否存在多个表示方法但是表示同一个国家的数据情况

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


我们发现U.K UK都代表了United Kingdom ，而United States 则代表了USA 后续要进行替代

### 评估无效或错误数据

这里使用desrible来获得dataframe的对数值统计消息快速了解

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

此时发现 CustomerID也被计算了，这个应当是字符串，之后要进行修改。同时发现交易数量Quantity与单价UnitPrice出现了最小值为负数的异常情况，要对此进行进一步评估

In [11]:
original_data[original_data['Quantity']<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，Quantity 为负数，InvoiceNo似乎以C开头，表示订单取消，追加条件，是否存在Quantity为负数，但是InvoiceNo不以C开头的观察值

In [12]:
original_data[(original_data['Quantity']<0) & (original_data['InvoiceNo'].str[0]!='C') ]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

Quantity为负数，同时存在InvoiceNo不以C开头的观察值。
但是根据以上输出结果可以得知，这些筛选出观察值为UnitPrice观察值均为0，因此增加UnitPrice的条件进行验证。

In [13]:
original_data[(original_data['Quantity']<0) & (original_data['InvoiceNo'].str[0]!='C') &(original_data['UnitPrice']!=0) ]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

观察后发现，Quantity为负数，同时存在InvoiceNo不以C开头的观察值，此时UnitPrice观察值均为0，这些数据均不贡献销售，所以之后可以进行清除

接下来筛选UnitPrice为负数的观察值

In [14]:
original_data[original_data['UnitPrice']<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

此时发现UnitPrice为负数的观察值为坏账调整，不属于实际交易，所以可以在清理数据中删除。

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括，
把 InvoiceDate 变量的数据类型转换为为日期时间
把 CustomerID 变量的数据类型转换为字符串
把 Description 变量缺失的观察值删除
把 Country 变量值"USA"替换为 "United States
把
,Country 变量值"UK"、"U.K."替换为"United Kingdom'
把 Quantity 变量值为负数的观察值删除
把 UnitPrice 变量值为负数的观察值删除

In [ ]:
#用copy,复制一个原本，并赋值。

In [15]:
clean_data = original_data.copy()

In [25]:
#把InvoiceDate转化为日期。
clean_data['InvoiceDate'] = pd.to_datetime(clean_data["InvoiceDate"])
clean_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

In [23]:
#把customerID转化成字符串
clean_data['CustomerID'] = clean_data["CustomerID"].astype(str)
clean_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

去除一下.0结尾

In [32]:
clean_data["CustomerID"] = clean_data['CustomerID'].str.slice(0,5)
clean_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把 Description 变量缺失的观察值删除,并查看删除后该列的个数和

In [33]:
clean_data.dropna(subset=['Description'],inplace= True)

In [34]:
clean_data["Description"].isnull().sum()

0

把 Country 变量值"USA"替换为 "United States

In [35]:
clean_data['Country'] = clean_data['Country'].replace({"USA":"Untiter States"})

In [36]:
len(clean_data[clean_data["Country"] == "USA"])

0

Country 变量值"UK"、"U.K."替换为"United Kingdom'

In [37]:
clean_data['Country'] = clean_data['Country'].replace({"UK":"United Kingdom"})
clean_data['Country'] = clean_data['Country'].replace({"U.K.":"United Kingdom"})

In [39]:
len(clean_data[clean_data["Country"] == "UK"])

0

In [40]:
len(clean_data[clean_data["Country"] == "U.K."])

0

把 Quantity 变量值为负数的观察值删除,并检查替换后的结果

In [41]:
#筛选出Quantity＞0的值，并将其再次复制给clean_data完成对Quantity为负数的观察值删除
clean_data = clean_data[clean_data["Quantity"] >= 0 ]

In [42]:
len(clean_data[clean_data["Quantity"] < 0 ])

0

把 UnitPrice 变量值为负数的观察值删除,并检查替换后的结果

In [45]:
clean_data = clean_data[clean_data["UnitPrice"] >= 0 ]

In [46]:
len(clean_data[clean_data["UnitPrice"] < 0 ])

0

## 保存清理后的数据

In [50]:
clean_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [52]:
#如果用to_csv保存，会发现前面有索引，这个索引我们不需要，所以用Index= Flase去取消索引，
#此时已完成清理，进行保存和后续分析。
clean_data.to_csv("e_commerce_cleaned.csv",index = False)

In [53]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\Luccklimt\AppData\Local\Temp\ipykernel_25088\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom